viernullvier/backend-coding-challenge-django
# feat(notes): Introduce public notes

## Scope

- Introduce new `public` flag on notes
- Public notes can be accessed by all users (authenticated and unauthenticated), but only edited and deleted by the author
- Notes can be filtered by their `public` status

## Out of scope

- It might be a bit unintuitive that the `/notes/` endpoint will now start to include notes by other users. If users have been expecting to see only their own nodes, the requirements would need to be specified further.
- It's currently not possible to see the public notes of just a single user. This would require another filter that hasn't been implemented as part of this PR.
- Test cases for this feature are relying on insertion order and will break if ordering is introduced. A better way to do this would be an order-independent assertion.

## How to test / reproduce

1. Create a new note or edit an existing one with the `public` flag set to `false`
2. Log out and call the `/notes/` endpoint
3. You should now be able to see public notes, but not edit or delete them
4. Log into a different user account than the one used in 1. and call the `/notes/` endpoint
5. You should now be able to see the note created in 1. alongside your notes

## Scouting 🏕️

- Permissions on the `/users/` endpoint have been cleaned up/fixed as well. Users should no longer be able to modify other users' accounts through the API.

## Review checklist

- [x] PR is ready
    - PR is split into meaningful commits (if needed) for the ease of reviewing
    - Description contains clear instructions on how to test the feature
- [x] Appropriate labels have been applied
